### PR TITLE
Dropdown, menus and popups

### PR DESCRIPTION
**Pull request by devkanro**
截屏2023-06-19 02 03 38
截屏2023-06-19 02 03 48

### CODE REVIEW

File: compose-utils/build.gradle.kts [hunk 1]
```diff
@@ -7,7 +7,7 @@ plugins {
 
 dependencies {
     api(compose.desktop.currentOs) {
-        exclude(group = "org.jetbrains.compose.material")
+        //exclude(group = "org.jetbrains.compose.material")
```
[the remark this answers is not in the document]
Comment (devkanro) [reply compose-utils/build.gradle.kts:10]: This is an erroneous commit, I annotated it temporarily for easy access to the material's code

I will remove it later

File: core/src/main/kotlin/org/jetbrains/jewel/internal/ContextMenu.kt [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+    Popup(
+        focusable = focusable,
+        onDismissRequest = onDismissRequest,
+        popupPositionProvider = rememberCursorPositionProvider(offset),
+        onKeyEvent = {
+            handlePopupMenuOnKeyEvent(it, focusManager!!, inputModeManager!!, menuManager)
+        }
+    ) {
+        focusManager = LocalFocusManager.current
```
[the remark this answers is not in the document]
Comment (devkanro) [reply core/src/main/kotlin/org/jetbrains/jewel/internal/ContextMenu.kt:64]: This code referred the code of material, and I don't to specifically check the reason, but I think the popup creates a new SkiaOwner, and the InputModeManager and FocusManager may not be related to the external

File: core/src/main/kotlin/org/jetbrains/jewel/internal/Link.kt [hunk 3]
```diff
@@ -135,30 +134,48 @@ fun DropdownLink(
...
-    textAlign = textAlign,
-    overflow = overflow,
-    lineHeight = lineHeight,
-    interactionSource = interactionSource,
-    indication = indication,
-    defaults = defaults,
-    colors = colors
+    colors: LinkColors = defaults.colors(),
+    menuModifier: Modifier = Modifier,
+    menuDefaults: MenuDefaults = IntelliJTheme.dropdownDefaults,
```
[the remark this answers is not in the document]
Comment (devkanro) [reply core/src/main/kotlin/org/jetbrains/jewel/internal/Link.kt:139]: DropdownLink using the DropdownDefaults as their MenuDefaults, it make menu style same as dropdown control

File: core/src/main/kotlin/org/jetbrains/jewel/internal/Dropdown.kt [hunk 4]
```diff
@@ -0,0 +1,396 @@
...
+                content = menuContent
+            )
+        }
+    }
+}
+
+@Composable
+internal fun DropdownMenu(
+    onDismissRequest: () -> Unit,
+    focusable: Boolean = true,
```
[the remark this answers is not in the document]
Comment (devkanro) [reply core/src/main/kotlin/org/jetbrains/jewel/internal/Dropdown.kt:138]: This cod referred the code of material too, but I think we can remove this pram too, is should be always true

File: core/src/main/kotlin/org/jetbrains/jewel/internal/Dropdown.kt [hunk 5]
```diff
@@ -0,0 +1,396 @@
...
+fun Dropdown(
+    modifier: Modifier = Modifier,
+    enabled: Boolean = true,
+    menuModifier: Modifier = Modifier,
+    error: Boolean = false,
+    interactionSource: MutableInteractionSource = remember { MutableInteractionSource() },
+    defaults: DropdownDefaults = IntelliJTheme.dropdownDefaults,
+    colors: DropdownColors = defaults.colors(),
+    content: @Composable RowScope.() -> Unit,
+    menuContent: MenuScope.() -> Unit
```
[the remark this answers is not in the document]
Comment (devkanro) [reply core/src/main/kotlin/org/jetbrains/jewel/internal/Dropdown.kt:59]: The content lambda is simple, but menuContent is complex, I tend to have complex lambdas as the last parameter

File: core/src/main/kotlin/org/jetbrains/jewel/internal/Dropdown.kt [hunk 6]
```diff
@@ -0,0 +1,396 @@
...
+                    is FocusInteraction.Focus -> dropdownState = dropdownState.copy(focused = true)
+                    is FocusInteraction.Unfocus -> dropdownState = dropdownState.copy(focused = false)
+                }
+            }
+        }
+
+        Box(
+            modifier.clickable(
+                onClick = {
+                    expanded = true
```
[the remark this answers is not in the document]
Comment (devkanro) [reply core/src/main/kotlin/org/jetbrains/jewel/internal/Dropdown.kt]: @rock3r  toggleable need a boolean state, we can't provide it for dropdown

File: core/src/main/kotlin/org/jetbrains/jewel/internal/Dropdown.kt [hunk 7]
```diff
@@ -85,7 +87,11 @@ fun Dropdown(
         Box(
             modifier.clickable(
                 onClick = {
-                    expanded = true
+                    // TODO: Trick to skip click event when close menu by click dropdown
```
Comment (rock3r) [inline core/src/main/kotlin/org/jetbrains/jewel/internal/Dropdown.kt:90]: One day we'll need a proper fix for this 🙃

**Review by rock3r — APPROVED** [review]

File: core/src/main/kotlin/org/jetbrains/jewel/internal/Dropdown.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (rock3r) [reply core/src/main/kotlin/org/jetbrains/jewel/internal/Dropdown.kt:60]: By the linked style guide, the composable content should always be the last parameter... I don't have enormously strong feelings on this, but I'd appreciate the extra bit of consistency

**Comment by rock3r** [comment]
@devkanro I'm ok with shipping this as it is, up to you if you wanna act on the last comment. If not just mark it as resolved.

**Review by fscarponi — APPROVED** [review]

**Pull request merged**